### PR TITLE
Subevent list: Add meta-data filter (Z#23114466)

### PR DESCRIPTION
**Pull request by wiffbi**
This PR adds meta-data filters to the dates/subevents overview.

In the second commit, I added some TODOs/questions in code comments.

### CODE REVIEW

File: src/pretix/control/forms/filter.py [hunk 1]
```diff
@@ -1116,9 +1116,31 @@ class SubEventFilterForm(FilterForm):
...
+                label=p.name,
+                required=False,
+                widget=forms.TextInput(
+                    attrs={
+                        'data-typeahead-url': reverse('control:event.subevents.meta.typeahead', kwargs={
+                            'organizer': self.organizer.slug,
+                            'event': self.event.slug
+                        }) + '?' + urlencode({
+                            'property': p.name,
+                            'organizer': self.organizer.slug if self.organizer else ''
```
Comment (raphaelm) [inline src/pretix/control/forms/filter.py:1139]:
```suggestion
                            'organizer': self.organizer.slug if self.organizer else ''
```
pointless if the typeahead url is event-scoped

File: src/pretix/control/forms/filter.py [hunk 2]
```diff
@@ -1181,13 +1203,49 @@ def filter_qs(self, qs):
...
+                    property__pk=p.pk,
+                )
+                qs = qs.annotate(**{'attr_{}'.format(i): Exists(semv_with_value)})
+                if p.name in filters_by_property_name:
+                    filters_by_property_name[p.name] |= Q(**{'attr_{}'.format(i): True})
+                else:
+                    filters_by_property_name[p.name] = Q(**{'attr_{}'.format(i): True})
+                if p.default == d:
+                    qs = qs.annotate(**{'attr_{}_any'.format(i): Exists(semv_with_any_value)})
+                    filters_by_property_name[p.name] |= Q(**{'attr_{}_any'.format(i): False, 'organizer_id': p.organizer_id})
```
Comment (raphaelm) [inline src/pretix/control/forms/filter.py:1226]:
```suggestion
                    filters_by_property_name[p.name] |= Q(**{'attr_{}_any'.format(i): False})
```

File: src/pretix/control/forms/filter.py [hunk 3]
```diff
@@ -1181,13 +1203,49 @@ def filter_qs(self, qs):
...
+        if self.organizer:
+            return self.organizer.meta_properties.filter(filter_allowed=True)
+        else:
+            # We ignore superuser permissions here. This is intentional – we do not want to show super
+            # users a form with all meta properties ever assigned.
+            return EventMetaProperty.objects.filter(
+                organizer_id__in=self.request.user.teams.values_list('organizer', flat=True),
+                filter_allowed=True,
+            )
+
```
Comment (raphaelm) [inline src/pretix/control/forms/filter.py:1248]:
```suggestion
        return self.organizer.meta_properties.filter(filter_allowed=True)
```

File: src/pretix/control/views/typeahead.py [hunk 4]
```diff
@@ -743,6 +744,44 @@ def meta_values(request):
...
+
+    organizer = get_object_or_404(Organizer, slug=organizer)
+    if not request.user.has_organizer_permission(organizer, request=request):
+        raise PermissionDenied()
+
+    event = get_object_or_404(Event, slug=event)
+    # TODO: do we need to check read permissions on the event here?
+
+    defaults = defaults.filter(organizer_id=organizer.pk)
+    matches = matches.filter(subevent__event_id=event.pk)
```
Comment (raphaelm) [inline src/pretix/control/views/typeahead.py:768]:
```suggestion
    defaults = defaults.filter(organizer_id=request.organizer.pk)
    matches = matches.filter(subevent__event_id=request.event.pk)
```

File: src/pretix/control/views/typeahead.py [hunk 5]
```diff
@@ -743,6 +744,44 @@ def meta_values(request):
...
+    # TODO: we anyway filter on one event, so we do not filter down to allowed events - see above if we need to check permissions on event
+#    all_access = (
+#        (request and request.user.has_active_staff_session(request.session.session_key))
+#        or request.user.teams.filter(all_events=True, organizer=organizer).exists()
+#    )
+#    if not all_access:
+#        matches = matches.filter(subevent__event__id__in=request.user.teams.values_list('limit_events__id', flat=True))
```
Comment (raphaelm) [inline src/pretix/control/views/typeahead.py:775]:
```suggestion
```

**Review by raphaelm — CHANGES_REQUESTED** [review]

File: src/pretix/control/forms/filter.py (hunk 1 above)
Comment (wiffbi) [reply src/pretix/control/forms/filter.py:1139]: I think you meant removing the if-else-part? That is right, pointless. I will remove it.

Comment (raphaelm) [reply src/pretix/control/forms/filter.py]: No, removing the entire line. Why do you need the organizer as the query argument when it's already in the URL?

Comment (wiffbi) [reply src/pretix/control/forms/filter.py]: Oh, sorry! Totally right.

**Review by raphaelm — APPROVED** [review]

**Pull request merged**